src/OpenTelemetry.Instrumentation.AWSLambda/Implementation/AWSMessagingUtils.cs

```diff
@@ -0,0 +1,137 @@
+// <copyright file="AWSMessagingUtils.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Amazon.Lambda.SNSEvents;
+using Amazon.Lambda.SQSEvents;
+using Newtonsoft.Json;
+using OpenTelemetry.Context.Propagation;
+
+namespace OpenTelemetry.Instrumentation.AWSLambda.Implementation;
+
+internal class AWSMessagingUtils
+{
+    // SNS attribute types: https://docs.aws.amazon.com/sns/latest/dg/sns-message-attributes.html
+    private const string SnsAttributeTypeString = "String";
+    private const string SnsAttributeTypeStringArray = "String.Array";
+    private const string SnsMessageAttributes = "MessageAttributes";
+
+    internal static PropagationContext ExtractParentContext(SQSEvent sqsEvent)
+    {
+        // We assume there can be only one parent that's why we consider only a single (the last) record as the carrier.
+        var message = sqsEvent?.Records?.LastOrDefault();
+        return ExtractParentContext(message);
+    }
+
+    internal static PropagationContext ExtractParentContext(SQSEvent.SQSMessage sqsMessage)
+    {
```

Comment:
Shouldn't the parent be the trace context created by the Lambda function?

Also, shouldn't we be creating span links to the trace contexts in the MessageAttributes?

Reply:
As above: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1051#discussion_r1147241174


This is the parent *of* the span created by the Lambda function.

Reply:
It looks like we are using the span context in the message attribute as the parent. I believe this span context should be a span link, not the parent.

What happens in the fan-in case? Say there are multiple messages in the Lambda event that are using in a downstream call?

Reply:
I think you are right and we should add links for all messages as well. However, I do think that adding one of the messages as parent (in this case the last one) is useful as well and should stay.

Reply:
I added links for SQS messages. Please, see the commit: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1051/commits/e88edbe98534c65924834d96edc2b0b1a706fb8a

Reply:
@atshaw43 Your concern should already be addressed in the newest version of this PR, as @rypdal added an option to configure whether a parent should be set or not (if not, the current context, if any, will be the implicit parent). See https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1051#issuecomment-1505516804 and the diff since then (https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1051/files/24ae92dfdcdd1ed8914bce51f08698fb41ece7a5..0e4d3857c649b9c4f99570d1103fabd37b697779).

Also note that the typical usage context will be different here than in your Java PR. In this PR here, we know that we are inside an AWS Lambda, so there typically will not be any implicit parent, the Lambda invocation span that we create here will be the (local) root span. Still, if there would be any parent span, I think setting the SetParentFromMessageBatch configuration to false (which is even the default now) should use it.

Reply:
As @rypdal made me aware, we actually never use the implicit / current context as parent, contrary to my previous comment. But this applies also to the existing old HTTP code paths. I think adjusting this should be done in a follow-up. You do have the possibility to pass in the current context explicitly when calling this library's API though.

Reply:
We may discuss using of the implicit / current parent in the follow-up PR. It may trigger further discussions in this already long playing PR.